libr/anal/esil_dfg.c

```diff
@@ -26,12 +30,439 @@ void _dfg_node_free (RAnalEsilDFGNode *free_me) {
 	free (free_me);
 }
 
-static bool _edf_reg_set(RAnalEsilDFG *edf, const char *reg, RGraphNode *node) {
-	return edf ? !sdb_ptr_set (edf->latest_nodes, reg, node, 0) : false;
+typedef struct esil_dfg_reg_var_t {
+	ut32 from;
+	ut32 to;
+	RGraphNode *node;
+} EsilDFGRegVar;
+
+
+static int _rv_del_alloc_cmp (void *incoming, void *in, void *user) {
+	EsilDFGRegVar *rv_incoming = (EsilDFGRegVar *)incoming;
+	EsilDFGRegVar *rv_in = (EsilDFGRegVar *)in;
+#if DFG_DEBUG
+	eprintf ("del_alloc_cmp:\n\tincoming: [%d-%d]\n\tin: [%d-%d]\n", rv_incoming->from, rv_incoming->to, rv_in->from, rv_in->to);
+#endif
+
+// first handle the simple cases without intersection
+	if (rv_incoming->to < rv_in->from) {
+#if DFG_DEBUG
+		eprintf ("del_alloc_cmp: case 0 - return -1\n");
+#endif
+		return -1;
+	}
+	if (rv_in->to < rv_incoming->from) {
+#if DFG_DEBUG
+		eprintf ("del_alloc_cmp: case 1 - return 1\n");
+#endif
+		return 1;
+	}
+	if (rv_in->from == rv_incoming->from && rv_in->to == rv_incoming->to) {
+#if DFG_DEBUG
+		eprintf ("del_alloc_cmp: case 2 - return 0\n");
+#endif
+		return 0;
+	}
+	RAnalEsilDFG *dfg = (RAnalEsilDFG *)user;
+
+/*
+the following cases are about intersection, here some ascii-art, so you understand what I do
+
+     =incoming=
+=========in=========
+
+split in into 2 and reinsert the second half (in2)
+shrink first half (in1)
+
+     =incoming=
+=in1=          =in2=
+
+*/
+
+	if (rv_in->from < rv_incoming->from && rv_incoming->to < rv_in->to) {
+#if DFG_DEBUG
+		eprintf ("del_alloc_cmp: case 3 - return 1\n");
+#endif
+		EsilDFGRegVar *rv = R_NEW (EsilDFGRegVar);
+		rv[0] = rv_in[0];
+		rv_in->to = rv_incoming->from - 1;
+		rv->from = rv_incoming->to + 1;
+		dfg->insert = rv;
+		return 1;
+	}
+
+/*
+   =incoming=
+      =in=
+
+enqueue the non-intersecting ends in the todo-queue
+*/
+
+	if (rv_incoming->from < rv_in->from && rv_in->to < rv_incoming->to) {
+#if DFG_DEBUG
+		eprintf ("del_alloc_cmp: case 4 - return 0\n");
+#endif
+		// lower part
+		EsilDFGRegVar *rv = R_NEW (EsilDFGRegVar);
+		rv[0] = rv_incoming[0];
+		rv->to = rv_in->from - 1;
+		r_queue_enqueue (dfg->todo, rv);
+		// upper part
+		rv = R_NEW (EsilDFGRegVar);
+		rv[0] = rv_incoming[0];
+		rv->from = rv_in->to + 1;
+		r_queue_enqueue (dfg->todo, rv);
+		return 0;
+	}
+
+/*
```

Comment:
kill those ugly commments

Reply:
why? they explain what the callback is doing, and how it is doing it.